<h3>Data Analytics II</h3>
<hr>
<p>1. Implement logisitic regression using Python to perform classification on Social_Network_Ads.csv dataset</p>
<p>2. Compute Confusion matrix to find TP, FP, TN, FN, Accuracy, Error rate, Precision, Recall on the given dataset.</p> 
<p>
<b>Name:</b> Udayan Chavan
<b>Roll no:</b> 31117
<b>Batch:</b> K1
</p>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')

In [2]:
data = pd.read_csv('Social_Network_Ads.csv')
data.head()

User ID  Gender  Age  EstimatedSalary  Purchased
0  15624510    Male   19            19000          0
1  15810944    Male   35            20000          0
2  15668575  Female   26            43000          0
3  15603246  Female   27            57000          0
4  15804002    Male   19            76000          0

In [3]:
data.describe()

User ID         Age  EstimatedSalary   Purchased
count  4.000000e+02  400.000000       400.000000  400.000000
mean   1.569154e+07   37.655000     69742.500000    0.357500
std    7.165832e+04   10.482877     34096.960282    0.479864
min    1.556669e+07   18.000000     15000.000000    0.000000
25%    1.562676e+07   29.750000     43000.000000    0.000000
50%    1.569434e+07   37.000000     70000.000000    0.000000
75%    1.575036e+07   46.000000     88000.000000    1.000000
max    1.581524e+07   60.000000    150000.000000    1.000000

Age and Salary are the factors to be considered. Extracting X (independent variable) and Y (dependent variable) from the dataset, taking third and fourth columns for X and fifth column for Y, we get:

In [4]:
X = data.iloc[:, [2,3]].values # contains age and salary data
Y = data.iloc[:, 4].values # contains ad click data

print("Age and Salary data:")
print(X[:5])

print("="*15)

print("Ad click data:")
print(Y[:5])

Age and Salary data:
[[   19 19000]
 [   35 20000]
 [   26 43000]
 [   27 57000]
 [   19 76000]]
Ad click data:
[0 0 0 0 0]


Splitting the dataset into training and testing datasets, we get:

In [5]:
from sklearn.model_selection import train_test_split

# Splitting dataset into test data (25%) and train data (75%)
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.25, random_state = 0)

print("Training data X: ")
print(X_train[:5])
print("-"*16)

print("Training data Y: ")
print(Y_train[:5])
print("\n================")

print("\nTesting data X: ")
print(X_test[:5])
print("-"*16)

print("Testing data Y: ")
print(Y_test[:5])

Training data X: 
[[    44  39000]
 [    32 120000]
 [    38  50000]
 [    32 135000]
 [    52  21000]]
----------------
Training data Y: 
[0 1 0 1 1]


Testing data X: 
[[   30 87000]
 [   38 50000]
 [   35 75000]
 [   30 79000]
 [   35 50000]]
----------------
Testing data Y: 
[0 0 0 0 0]


Applying feature scaling to normalize data, we get:

In [6]:
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()

X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

print(X_train[:5])
print("-"*28)
print(X_test[:5])

[[ 0.58164944 -0.88670699]
 [-0.60673761  1.46173768]
 [-0.01254409 -0.5677824 ]
 [-0.60673761  1.89663484]
 [ 1.37390747 -1.40858358]]
----------------------------
[[-0.80480212  0.50496393]
 [-0.01254409 -0.5677824 ]
 [-0.30964085  0.1570462 ]
 [-0.80480212  0.27301877]
 [-0.30964085 -0.5677824 ]]


Importing Logisitic regression class from sklearn library, we get:

In [7]:
from sklearn.linear_model import LogisticRegression

classifier = LogisticRegression(random_state = 0, solver = "lbfgs")

# Fitting the classifier to the training set
classifier.fit(X_train, Y_train)

# Using .predict() to get a vector of predictions of X_test
Y_pred = classifier.predict(X_test)

# Printing predicted data
print(X_test[:10])
print("-"*28)
print(Y_pred[:10])

[[-0.80480212  0.50496393]
 [-0.01254409 -0.5677824 ]
 [-0.30964085  0.1570462 ]
 [-0.80480212  0.27301877]
 [-0.30964085 -0.5677824 ]
 [-1.10189888 -1.43757673]
 [-0.70576986 -1.58254245]
 [-0.21060859  2.15757314]
 [-1.99318916 -0.04590581]
 [ 0.8787462  -0.77073441]]
----------------------------
[0 0 0 0 0 0 0 1 0 1]


<p> Observing the predicted data, we can see that as the value of independent variables (age and salary) increases, the user is more likely to click on an ad.</p><hr>
Comparing predicted data to test data:

In [8]:
print("Predicted data:", Y_pred[:25])
print("Test data:     ", Y_test[:25])

Predicted data: [0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 1 0 0 1 0 1 0]
Test data:      [0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 1 0]


In the first 25 entries, only one mismatch is found. Accuracy of the model is good. Now, using confusion matrix to evaluate how accurate our model is, we get:

In [9]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, Y_pred)

print(cm)

[[65  3]
 [ 8 24]]


The sum of top-left and bottom-right values gives us the number of of correct predictions, and the other two values give us the number of incorrect predictions. 
<p>
In our case, we have 89 (65 + 24) correct predictions and 11 (8 + 3) incorrect predictions.
</p>

<p>
<b>TP: True positives </b> Cases that we predicted true but are actually true. (Bottom right value in confusion matrix): <b>24</b>
</p>

<p>
<b>FP: False positives </b> Cases that we predicted true but are actually false. (Top right value in confusion matrix): <b>3</b>
</p>

<p>
<b>TN: True negatives </b> Cases that we predicted false but are actually false. (Top left value in confusion matrix): <b>65</b>
</p>

<p>
<b>FN: False negatives </b> Cases that we predicted false but are actually true. (Bottom left value in confusion matrix): <b>8</b>
</p>

<hr>

<p>
<b>Accuracy: </b> (true positives + true negatives) / (true positives + true negatives + false positives + false negatives): (24 + 65) / 100 = <b>0.89</b>
</p>

<p>
<b>Error rate: </b> (false positives + false negatives) / (true positives + true negatives + false positives + false negatives): (3 + 8) / 100 = <b>0.11</b>
</p>

<p>
<b>Precision: </b> true positives / (true positives + false positives): 24 / (24 + 3) = <b>0.889</b>
</p>

<p>
<b>Recall: </b> true positives / (true positives + false negatives): 24 / (24 + 8) = <b>0.75</b>
</p>